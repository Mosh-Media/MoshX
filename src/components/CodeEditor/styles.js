const styles = theme => ({
	editor: {
		transition: "all 500ms",
		font:
			"16px/normal 'Monaco', 'Menlo', 'Ubuntu Mono', 'Consolas', 'source-code-pro', monospace",
		[theme.breakpoints.up(600)]: {
			width: "380",
			marginLeft: "auto",
			marginRight: "auto"
		},
		"& textarea": {
			fontSize: "14px",
			outline: "none"
		}
	},
	preview: {
		width: "300px",
		margin: "0 auto",
		display: "flex",
		justifyContent: "center",
		borderRadius: "8px",
		[theme.breakpoints.up(600)]: {
			width: "300px",
		}
	},
	card: {
		background: "white",
		borderRadius: "8px",
		boxShadow:
			"0 24px 24px -18px rgba(69,104,129,.33), 0 9px 45px 0 rgba(114,119,160,.12)",
		padding: "1.2em"
	},
	error: {
		background: "rgb(255, 85, 85)",
		color: "rgb(248, 248, 242)",
		padding: "10px",
		borderRadius: "8px",
		boxShadow:
			"0 24px 24px -18px rgba(69,104,129,.33), 0 9px 45px 0 rgba(114,119,160,.12)"
	},
	correct: {
		boxShadow:
			"0 0 114px 0 rgba(0,0,0,.08), 0 30px 25px 0 rgba(196, 245, 46, 0.49)"
	}
});

export default styles;
