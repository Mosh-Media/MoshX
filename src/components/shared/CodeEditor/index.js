import React, { lazy, useState, useEffect, Suspense } from "react";

import Grid from "@material-ui/core/Grid";

const LiveProviderCore = lazy(() => import("./LiveProviderCore"));

const CodeEditor = ({ handleOnChange, md, sm, matchPercent, question }) => {
  const [state, setState] = useState(question);

  useEffect(() => {
    if (question) {
      try {
        /* Questions can contain special JSON characters that needs to be parsed */
        const parseQuestion = JSON.parse(question.question);
        setState({ ...question, question: parseQuestion });
      } catch {
        setState(question);
      }
    }
  }, [question]);

  return (
    <Grid container spacing={4}>
      <Suspense>
        <LiveProviderCore
          question={state}
          handleOnChange={handleOnChange}
          md={md}
          sm={sm}
          matchPercent={matchPercent}
        />
      </Suspense>
    </Grid>
  );
};

export default CodeEditor;
