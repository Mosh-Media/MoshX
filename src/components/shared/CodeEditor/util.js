/* * *  TODO  * * * * * * * * * * * * * * * * * * * * * *
 *  Figure out a new way to do this                    *
 * * * * * * * * * * * * * * * * * * * * * * * * * * */

export const addFocusOnEditor = async () => {
  await wait(400);
  try {
    document.querySelector(".npm__react-simple-code-editor__textarea").focus();
  } catch {
    console.log("Focus on editor failed");
  }
};

export const getPreviewElement = () => {
  let element;
  try {
    element = document.querySelector("#MoBrowserMockup > div");
  } catch {
    console.log("getPreviewElement not available");
  }
  return element;
};

async function wait(ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms);
  });
}
