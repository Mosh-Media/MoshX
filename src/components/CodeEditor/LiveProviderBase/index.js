import React from "react";

import { LiveEditor, LiveProvider, LivePreview, LiveError } from "react-live";
import Grid from "@material-ui/core/Grid";
import Headline from "../../shared/Headline";
import styles from "./styles";
import Slide from "@material-ui/core/Slide";
import Title from "../../../components/shared/Title";
import withStyles from "@material-ui/core/styles/withStyles";
import { reactLiveTheme } from "../../../utils/reactLiveTheme";

const LiveProviderBase = ({ classes, question, handleOnChange }) => {
	return (
			<LiveProvider code={question.question} language="jsx" noInline={false}>
				<Grid item md={6} sm={12} xs={12} style={{ width: "100%" }}>
					<Title text={question.label} fade={question.label && true} />
					<Slide
						direction="right"
						in={question.question && true}
						mountOnEnter
						timeout={{ enter: 800, exit: 400 }}
						unmountOnExit
					>
						<div
							className={`${classes.editor} ${classes.card} ${question.isCorrect &&
								classes.correct}`}
						>
							<LiveEditor onChange={handleOnChange} theme={reactLiveTheme} />
						</div>
					</Slide>
				</Grid>
				<Grid item md={6} sm={12} xs={12}>
					<Title text="Code Preview" fade={true} />
					<Slide
						direction="left"
						in={(question.isPlayground && true) || (question.question && true)}
						mountOnEnter
						timeout={{ enter: 800, exit: 800 }}
						unmountOnExit
					>
						<div className={classes.browserMockup}>
							<div className={classes.browserButtons}></div>
							<div className={classes.browserButtons2}></div>
							<div className={classes.preview} >
							<LivePreview />
							</div>
						</div>
					</Slide>
				</Grid>
				{question.isPlayground && (
					<Grid item md={12} sm={12} xs={12}>
						<LiveError />
					</Grid>
				)}
				<Grid item md={12} sm={12} xs={12}>
					<Headline isCorrect={question && question.isCorrect} />
				</Grid>
			</LiveProvider>
	);
};

export default withStyles(styles)(LiveProviderBase);
