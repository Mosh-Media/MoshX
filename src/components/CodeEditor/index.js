import React from "react";

import { LiveEditor, LiveProvider, LivePreview, LiveError } from "react-live";
import { Grid } from "@material-ui/core";
import Headline from "../shared/Headline";
import styles from "./styles";
import Slide from "@material-ui/core/Slide";
import Title from "../../components/shared/Title";
import withStyles from "@material-ui/core/styles/withStyles";
import { reactLiveTheme } from "../../utils/reactLiveTheme";

const CodeEditor = ({ classes, question, handleOnChange }) => {
	return (
		<Grid container spacing={4}>
			<LiveProvider code={question.question} language="jsx" noInline={false}>
				<Grid item md={6} sm={12} xs={12} style={{ width: "100%" }}>
					<Title text={question.label} fade={question.label && true} />
					<Slide
						direction="right"
						in={question.label && true}
						mountOnEnter
						timeout={{ enter: 400, exit: 400 }}
						unmountOnExit
					>
						<div
							className={`${classes.editor} ${
								classes.card
							} ${question.isCorrect && classes.correct}`}
						>
							<LiveEditor onChange={handleOnChange} theme={reactLiveTheme} />
						</div>
					</Slide>
				</Grid>
				<Grid item md={6} sm={12} xs={12}>
					<Title text="Code Preview" fade={true} />
					<Slide
						direction="left"
						in={question.isPlayground || question.question}
						mountOnEnter
						timeout={{ enter: 400, exit: 400 }}
						unmountOnExit
					>
						<div className={classes.browserMockup}>
							<div className={classes.browserButtons}></div>
							<div className={classes.browserButtons2}></div>
							<LivePreview className={classes.preview} />
						</div>
					</Slide>
				</Grid>
				{question.isPlayground && (
					<Grid item md={12} sm={12} xs={12}>
						<LiveError />
					</Grid>
				)}
				<Grid item md={12} sm={12} xs={12}>
					<Headline isCorrect={question && question.isCorrect} />
				</Grid>
			</LiveProvider>
		</Grid>
	);
};

export default withStyles(styles)(CodeEditor);
