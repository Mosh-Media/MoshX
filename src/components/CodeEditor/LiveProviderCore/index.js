import React, { useEffect } from "react";

import { addFocusOnEditor } from "./util";
import { LiveEditor, LiveProvider, LivePreview, LiveError } from "react-live";
import Grid from "@material-ui/core/Grid";
import Headline from "../../shared/Headline";
import styles from "./styles";
import Slide from "@material-ui/core/Slide";
import MoTitle from "../../shared/MoTitle";
import withStyles from "@material-ui/core/styles/withStyles";
import { reactLiveTheme } from "../../../utils/reactLiveTheme";

const LiveProviderCore = ({ classes, question, handleOnChange }) => {
	useEffect(() => {
		addFocusOnEditor();
	}, []);

	return (
		<LiveProvider code={question.question} language="jsx" noInline={false}>
			<Grid item md={6} sm={12} xs={12} style={{ width: "100%" }}>
				<MoTitle
					text={question.label}
					fade={question.label && true}
					margin="36px 0 36px"
				/>
				<Slide
					direction="right"
					in={question.question && true}
					mountOnEnter
					timeout={{ enter: 800, exit: 400 }}
					unmountOnExit
				>
					<div
						className={`${classes.browserMockup} ${question.isCorrect &&
							classes.correct}`}
					>
						<div className={classes.browserButtons}></div>
						<div className={classes.browserFilename}>
							<span>{`index.${question.language}`}</span>
						</div>
						<div className={classes.editor}>
							<LiveEditor onChange={handleOnChange} theme={reactLiveTheme} />
						</div>
					</div>
				</Slide>
			</Grid>
			<Grid item md={6} sm={12} xs={12}>
				<MoTitle text="Live Preview" fade={true} margin="36px 0 36px" />
				<Slide
					direction="left"
					in={(question.isPlayground && true) || (question.question && true)}
					mountOnEnter
					timeout={{ enter: 800, exit: 800 }}
					unmountOnExit
				>
					<div className={classes.browserMockup}>
						<div className={classes.browserButtons}></div>
						<div className={classes.browserButtons2}>
							<span>localhost:3000</span>
						</div>
						<div className={classes.preview}>
							<LivePreview />
						</div>
					</div>
				</Slide>
			</Grid>
			{question.isPlayground && (
				<Grid item md={12} sm={12} xs={12}>
					<LiveError />
				</Grid>
			)}
			<Grid item md={12} sm={12} xs={12}>
				<Headline isCorrect={question && question.isCorrect} />
			</Grid>
		</LiveProvider>
	);
};

export default withStyles(styles)(LiveProviderCore);
