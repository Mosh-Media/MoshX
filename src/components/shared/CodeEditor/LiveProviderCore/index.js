import React, { useEffect } from "react";

import { addFocusOnEditor } from "./util";
import { LiveEditor, LiveProvider, LivePreview, LiveError } from "react-live";
import { reactLiveTheme } from "utils/reactLiveTheme";
import Grid from "@material-ui/core/Grid";
import Slide from "@material-ui/core/Slide";
import Headline from "components/library/MoHeadline";
import MoBrowserMockup from "components/library/MoBrowserMockup";
import MoLinearProgress from "components/library/MoLinearProgress";
import styles from "./styles";
import Typist from "react-typist";
import withStyles from "@material-ui/core/styles/withStyles";

const LiveProviderCore = ({
  classes,
  handleOnChange,
  md,
  sm,
  matchPercent,
  question,
}) => {
  useEffect(() => {
    addFocusOnEditor();
  }, [question]);

  return (
    <LiveProvider code={question.question} language="jsx" noInline={false}>
      <Grid item md={md} sm={sm} xs={12} style={{ width: "100%" }}>
        <Slide
          direction="right"
          in={question.answer && true}
          mountOnEnter
          timeout={{ enter: 400, exit: 400 }}
          unmountOnExit
        >
          <div>
            <MoBrowserMockup fileType={question.language} isEditor={true}>
              <LiveEditor onChange={handleOnChange} theme={reactLiveTheme} />
              {!question.question && (
                <Typist
                  className={classes.hint}
                  avgTypingDelay={60}
                  stdTypingDelay={30}
                  startDelay={800}
                  cursor={{
                    show: true,
                    blink: false,
                    element: (
                      <span
                        role="img"
                        aria-label="magical dust"
                        className={classes.typistEmoji}
                      >
                        âœ¨
                      </span>
                    ),
                    hideWhenDone: true,
                  }}
                >
                  {question.answer}
                </Typist>
              )}
              {matchPercent && (
                <MoLinearProgress percent={matchPercent * 100} />
              )}
            </MoBrowserMockup>
          </div>
        </Slide>
      </Grid>
      <Grid item md={md} sm={sm} xs={12}>
        <Slide
          direction="left"
          in={(question.isPlayground && true) || (question.answer && true)}
          mountOnEnter
          timeout={{ enter: 400, exit: 400 }}
          unmountOnExit
        >
          <div>
            <MoBrowserMockup fileType={question.language} isBrowser={true}>
              <LivePreview />
            </MoBrowserMockup>
          </div>
        </Slide>
      </Grid>
      {question.isPlayground && (
        <Grid item md={12} sm={12} xs={12}>
          <LiveError />
        </Grid>
      )}
      <Grid item md={12} sm={12} xs={12}>
        <Headline isCorrect={question && question.isCorrect} />
      </Grid>
    </LiveProvider>
  );
};

export default withStyles(styles)(LiveProviderCore);
