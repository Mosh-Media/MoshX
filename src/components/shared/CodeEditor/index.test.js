import React from 'react';

import ShallowRenderer from 'react-test-renderer/shallow';
import CodeEditor from './index.js';

const renderer = new ShallowRenderer();

describe('CodeEditor Component', () => {
  test('should match the snapshot', () => {
    renderer.render(<CodeEditor />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
