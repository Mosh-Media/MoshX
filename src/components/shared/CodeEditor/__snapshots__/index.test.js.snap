// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeEditor Component should match the snapshot 1`] = `
<ForwardRef(WithStyles)
  container={true}
  spacing={2}
>
  <LiveProvider
    code=""
    disabled={false}
    language="jsx"
    noInline={false}
  >
    <ForwardRef(WithStyles)
      item={true}
      style={
        Object {
          "width": "100%",
        }
      }
      xs={12}
    >
      <ForwardRef(Grow)
        mountOnEnter={true}
        timeout={
          Object {
            "enter": 600,
            "exit": 600,
          }
        }
        unmountOnExit={true}
      >
        <div>
          <ForwardRef(WithStyles)
            isEditor={true}
          >
            <div>
              <LiveEditor
                className="makeStyles-liveEditor-2"
                onChange={[Function]}
                theme={
                  Object {
                    "plain": Object {
                      "color": "#4D4D4C",
                    },
                    "styles": Array [
                      Object {
                        "style": Object {
                          "color": "hsl(30, 20%, 50%)",
                        },
                        "types": Array [
                          "prolog",
                          "comment",
                          "doctype",
                          "cdata",
                        ],
                      },
                      Object {
                        "style": Object {
                          "color": "#006080",
                        },
                        "types": Array [
                          "property",
                          "tag",
                          "boolean",
                          "number",
                          "constant",
                          "symbol",
                        ],
                      },
                      Object {
                        "style": Object {
                          "color": "#9c005d",
                        },
                        "types": Array [
                          "attr-name",
                          "string",
                          "char",
                          "builtin",
                          "insterted",
                        ],
                      },
                      Object {
                        "style": Object {
                          "color": "#065f00",
                        },
                        "types": Array [
                          "operator",
                          "entity",
                          "url",
                          "string",
                          "variable",
                          "language-css",
                        ],
                      },
                      Object {
                        "style": Object {
                          "color": "rgb(255, 85, 85)",
                        },
                        "types": Array [
                          "deleted",
                        ],
                      },
                      Object {
                        "style": Object {
                          "fontStyle": "italic",
                        },
                        "types": Array [
                          "italic",
                        ],
                      },
                      Object {
                        "style": Object {
                          "fontWeight": "bold",
                        },
                        "types": Array [
                          "important",
                          "bold",
                        ],
                      },
                      Object {
                        "style": Object {
                          "color": "#e90",
                        },
                        "types": Array [
                          "regex",
                          "important",
                        ],
                      },
                      Object {
                        "style": Object {
                          "color": "#006080",
                        },
                        "types": Array [
                          "atrule",
                          "attr-value",
                          "keyword",
                        ],
                      },
                      Object {
                        "style": Object {
                          "color": "black",
                          "opacity": "0.7",
                        },
                        "types": Array [
                          "punctuation",
                          "symbol",
                        ],
                      },
                    ],
                  }
                }
              />
            </div>
          </ForwardRef(WithStyles)>
        </div>
      </ForwardRef(Grow)>
    </ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      item={true}
      xs={12}
    >
      <ForwardRef(Grow)
        mountOnEnter={true}
        timeout={
          Object {
            "enter": 400,
          }
        }
        unmountOnExit={true}
      >
        <div>
          <ForwardRef(WithStyles)
            isEditor={false}
          />
        </div>
      </ForwardRef(Grow)>
    </ForwardRef(WithStyles)>
  </LiveProvider>
</ForwardRef(WithStyles)>
`;
